Instanced attributes sets

dedoardo opened the pull request:
**Description**
This patch allows objects to read from and index shared attribute sets.
The logic in the geometry manager has been changed to allocate an extra device attribute map for each `InstanceGroup` which is essentially an alias for an `AttributeSet`. The attribute requests are fulfilled by the instanced attribute set, or the geometry attribute set if the former are not found. 

Design considerations are very welcome. One drawback of the current approach is that geometry attributes which are overwritten by instanced attributes are currently committed to the device. Skipping this commit requires knowing which attributes in the geometry attribute set are *always* overwritten by instanced attribute sets, requiring extra management logic. It could also be argued that adding those attribute to the geometry in the first place doesn't make a lot of sense and maybe the responsibility lies in the delegate.

**Changes**
- Added a vector of attribute sets at the scene level which can be referenced by objects
- Changed `Geometry::device_update` to combine geometry and instanced attribute sets
- `find_attribute` uses the per object `instance_index` to index the attributeset
- The kernel object for point clouds has switched to using numverts, rather than numkeys to match the interpolation 
ATTR_ELEMENT_VERTEX

**Test**
Here is a test file 
[perpoint_attr.zip](https://github.com/tangent-opensource/coreBlackbird/files/6843559/perpoint_attr.zip)

![image](https://user-images.githubusercontent.com/2072636/126218557-8e9680d5-ef2d-4dc6-8326-abcba8950df4.png)

bareya commented on src/render/instance_group.h:
```diff
@@ -0,0 +1,43 @@
+class InstanceGroup {
+public:
+  // todo(Edo): Implement subdivision surfaces once they are internal
+  InstanceGroup(Geometry* geom, uint instances) : attributes(geom, ATTR_PRIM_GEOMETRY, instances), attr_map_offset(~0) { }
```
Why this is being initialized twice? one in the field to `0` and to `-1` in construction initialization list?

dedoardo commented on src/kernel/kernel_types.h:
```diff
@@ -1534,10 +1535,12 @@ typedef struct KernelObject {
 
   int numkeys;
-  int numverts;
+  int numverts; /* Number of vertices in a mesh or points in a cloud */
```
Yes, the goal is to override attributes with different interpolation (or element in cycles speak). If an attribute has `ATTR_ELEMENT_VERTEX` interpolation, we want to bump the offset by the number of vertices.
The current code was using numkeys for points in a cloud, while attributes authored on those point have ATTR_ELEMENT_VERTEX as interpolation. If we wanted to branch on the type of geometry to calculate the indexed offset, the code would have to be scattered and duplicated in the various `primitive_*_attribute`. I decided to make that change to streamline the logic.

So far numkeys/numverts was being used for motion blur positions, so faces wasn't needed until now. I know that on the delegate side we currently only support constants, but the USD specification doesn't impose this limit so I went for a more generic approach.

dedoardo replied:
If you are talking about the fact that numkeys and numverts are never set together and we could avoid using one of them, I totally agree. It hasn't been done in this PR to limit the number of changes.

dedoardo commented on src/render/attribute.h:
```diff
@@ -54,11 +54,15 @@ class Attribute {
   uint flags; /* enum AttributeFlag */
 
+  /* Multiplier on the size of the buffer */
+  uint instances;
```
Nope, the field can be made private. `Attribute::resize` already uses the number of instances to determine the buffer size. I don't think we need to expose the functionality you are talking about.

dedoardo commented on src/render/attribute.h:
```diff
@@ -175,8 +179,9 @@ class AttributeSet {
   Geometry *geometry;
   AttributePrimitive prim;
   list<Attribute> attributes;
+  uint instances;
```
They can't be different and the one in `Attribute` should be private. It's the attribute set that adds the attributes, but I forgot to patch the other `add` calls, will do that.

dedoardo commented:
Hi, thanks a lot for the review @bareya and @skwerner for the bugfix :) I made the following changes
- Removed double initialization of `InstanceGroup::attr_map_offset`
- Moved `Attribute::instances` to a private block and changed the type to size_t to be consistent with other  sizes
- Made `Geometry::num_points` virtual. There should probably be a simpler way to pass numverts/numpoints specific to each geometry and avoid duplication.
- More descriptive comments for internal variables

Maybe `Attribute::element_size` should include the number of instances. What do you think?
